### PR TITLE
Implement begingroup package and tests for it

### PR DESCRIPTION
**Pull request by dpvc**
This PR implements the `begingroup` package that was missing from v3, and creates tests for it.  It also adds two tests to the `newcommand` tests that completes the coverage for its files.

Some changes needed to be made to the `newcommand` package to make this work, and a few details were changed in other places as well.

* The `Configuration` priority handling was modified to allow a priority of 0 (not strictly necessary for this, but the fact that it didn't work before could cause confusion).

* New `remove()` methods are added to the `SubHandler` and `SubHandlers` classes in order to be able to remove the maps that are added when `\begingroup` is used.

* The handling of the `char === null` case for character maps that is used to make an "undefined" delimiter that masks one that is defined later in the map list has been moved from the `applicable()` to the `contains()` method, since we need to be able to locate the map that holds one of these definitions in order to be able to remove it when a global definition replaces it.

* The `ams` package now uses the `newcommand` maps rather than its own private map for `\DeclareMapOperator` definitions, so that they work with the `begingroup` infrastructure.

* Similarly, the `mathtools` paried-delimiter macros now use the `newcommand` maps as well.

* The `extpfeil` extension has always used the `newcommand` maps, but it now initialized the maps in a better way when newcommand isn't in the package list.

* The `newcommand` init functionality is now moved to an exported config function, so that it can create `packageData` where it can record whether the initialization has already been done.  The `ams`, `mathtools`, and `extpfeil` packages now call that config function to make sure the needed maps are in place.

* The `newcommand` maps are now given priority -100 rather than -1, since the `begingroup` and `newcommand` maps must be the first ones in the handler lists.  This gives a little more remove for other packages to insert maps between the default location and the `newcommand` ones.

* There is now a list of protected macro names that can't be redefined using `\def`, `\let`, `\newcommand`, or the other tools for creating macros (as long as they use the `newcommand` methods for doing so).  This allows sites like StackExchange to create sandboxed user regions that can't affect other user posts.  A new `\begingroupSandbox` macro is used to separate the sandboxes, and is the only macro on the protected list.

The other changes to `newcommand` are described below, and are there to make sure the definitions goes into the correct map (either the top-most `\begingroup` maps, or the base "global" maps for `\global\let`, `\gdef`, etc.).

The `begingroup` package defines a `BegingroupStack` object that manages the definition maps needed to isolate the definitions within the `\begingroup ... \endgroup` pairs.  A `\begingroup` call creates new macro, delimiter, and environment maps, and apends them to the handler list with a priority one less than the previous group (or the initial `newcommand` priority).  Since these have the same names as the `newcommand` maps, `newcommand` will put new definitions in the most recently created maps automatically.

An `\endgroup` causes the most recent set of maps to be removed (if there is one), and then re-registers the previous set of maps using the `newcommand` names.  That means `newcommand` will define to the correct maps again.

The only complication is how to handle `\global` definitions.  These need to be placed into the original `newcommand` maps, which are the mapos in effect when no `\begingroup` has been performed.  Those maps are stored in the `global` object of the `BegingroupStack` for easy reference.  In order to determine which maps to use, `begingroup` uses a value `isGlobal` in the `parser.stack.env` to track this.  When `newcommand` detects this value, it asks `begingroup` to give it the global maps rather than use its usual named ones.  The `begingroup` package stores its `BegingroupStack` instance in the parser's `packageData`, where `newcommand` can access it.  When `newcommand` calls the `begingroup`'s `checkGlobal()` function to get the global maps, that function also checks to see if the token about to be defined gloablly is defined in any of the stack maps, and removes those definitions so that the global version will be the one used when the token is accessed.  We do a little trickery here to be able to look through only the `begingroup` maps and not past them by placing a marker in the base global map that we detect (and remove), telling us to stop looking at that point.

The remainder of the changes to `newcommand` are basically just to switch to asking `begingroup` for the global maps when `isGlobal` is true, and to remove the `isGlobal` marker when the definition is made.

In addition to `\begingroup` and `\endgroup` (and `\global` and `\gdef`), the `begingroup` extension defines a `\begingroupReset` macro that removes all the `\begingroup` maps and resets to the base global maps.  

The `begingroup` package also defines `\begingroupSandbox` that can be used to isolate the definitions in one section of the page from those in another, so that sites like StackExchange can use this between user posts to make sure that one user doesn't redefine things to mess up another user.  The `\begingroupSandbox` macro can't be redefined, and its action is to remove all the begingroup maps and then push a fresh set (effectively doing `\begingroupReset\begingroup`).  This removes any previous user definitions and makes a new group for the next user's definition.  Then mark the fresh set of maps as the global maps.  This means that even if the math includes `\global` definitions, they will go into this new set of maps, not the actual global maps.  Those global maps will be replaced whent he next `\begingroupSandbox` is performed.  The original `newcommand` global maps can still be used to store shared definitions that are made before the first `\begingroupSandbox` call occurs.  Once `\begingroupSandbox` is performed, however, there is no going back; the original `newcommand` map are no longer accessible to new definitions.

The final thing to note is the pre- and post-filters that `begingroup` adds to the TeX input jax.  Because `\begingroup` can be performed in one expression and the `\endgroup` in another (so that you can have local definitions over several expressions), you need to be careful about how many groups are open when an expression starts so that if a retry occurs during the processing of an expression, say for loading a package via `\require` or by autoloading, you can remove the new group maps before re-processing the expression when the extension is loaded.

Consider the expression

``` latex
$\def\x{A}$ ... $\x \begingroup \def\x{B} \require{cancel} \cancel{\x+1} \endgroup$
```

The initial definition for the `\x` in the second expression is in the global map, and after the `\begingroup`, there is a new set of maps above the global maps; the second `\x` is defined there.  This is the state when `\require{}` is peformed, which causes a retry and eventually re-typesets the expression.  If we didn't do anything more, when the expression is typeset the second time (after `cancel` is loaded), the new maps are already in place, and the `\x` in the expression is found there, so it produces a `B`, not the expected `A`.  Then a *second* set of maps is pushed on the stack by the `\begingroup`, and after the `\endgroup`, we are still left with one begingroup set of maps on the stack that shouldn't be there.

To mitigate this, the pre-filter removes any maps that are above the recorded "top" of the stack, which initially is just at the base global maps.  The post-filter, which only runs of there are no retries, records the current stack height as the new top value.  In the senario above, the top is the global set when the second expression begins, and so nothing need to be cleared.  But after the retry, when it is re-typeset, the extra maps are first removed (since top is still the global set).  That resolves the problem from the previous paragraph.

``` latex
$\begingroup\def\x{A}$ ... $\x \begingroup \def\x{B} \require{cancel} \cancel{\x+1} \endgroup \x$ ... $\endgroup \x$
```

then when the first expression ends, there is one begingroup set of maps on the stack, so the top is at that value (-101 in this case, the priority of the top maps).  Then when the second expression is typeset, it adds a second group of maps, but when `\require` retries and the expression is retypeset, the second set is removed, but the first is retained.  At the end of that expression, we are back to the first begingroup maps again, and `\x` is `A`.  After the third expression, the begingroup maps are removed, and `\x` becomes undefined.

Anyway, this provides one of the big missing pieces from v2, with some additional functionality for use in multi-user settings like StackExchange.

### CODE REVIEW

**Review by zorkow — APPROVED** [review]
A number of minor things.
The only major part is that I believe we should break up the `BegingroupConfiguration.ts` file into three parts:
`BegingroupStack`, `BegingroupMethods` and `BegingroupConfiguration`. So far we tried (not always successful) to keep the actual configuration file small.

File: ts/input/tex/Configuration.ts [hunk 1]
```diff
@@ -84,7 +84,10 @@ export class Configuration {
       [ConfigurationType.PARSER]?: string;
     } = {}
   ): Configuration {
-    const priority = config.priority || PrioritizedList.DEFAULTPRIORITY;
+    const priority =
+      config.priority !== undefined
+        ? config.priority
+        : PrioritizedList.DEFAULTPRIORITY;
```
[the remark this answers is not in the document]
Comment (dpvc) [reply ts/input/tex/Configuration.ts]: done
